server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -614,4 +633,14 @@ public double getEwmaAlpha() {
         }
     }
 
+    public record QualifiedName(@Nullable String nodeName, String threadPoolName) {
+
+        public QualifiedName(String threadPoolName) {
+            this(null, threadPoolName);
+        }
+
+        public String toCompositeString() {
+            return nodeName == null ? threadPoolName : nodeName + "/" + threadPoolName;
+        }
+    }
```

Comment:
I added this so we could separate the node name and thread pool name for the purpose of generating valid metric names.

i.e. `es.thread_pool.instance-000003/write.threads.queue.latency.histogram` is invalid, we want `es.thread_pool.write.threads.queue.latency.histogram` instead, so better to pass in something structured and unambiguous than making assumptions about the format of the string.

Reply:
I wonder whether an alternative could be add a new method each to `ExecutorHolder` and `TaskExecutionTimeTrackingEsThreadPoolExecutor`, something like:

```java
class ExecutorHolder {
    void registerMetrics(MeterRegistry meterRegistry) {
        if (executor instanceof TaskExecutionTimeTrackingEsThreadPoolExecutor te) {
            te.registerMetrics(info.name, meterRegistry)
        }
    }
}

class TaskExecutionTimeTrackingEsThreadPoolExecutor {
    void registerMetrics(String name, MeterRegistry meterRegistry) {
        meterRegistry.registerLongsGauge(...);
    }
}
```

and call `ExecutorHolder#registerMetrics` after building it in `ThreadPool`. I think it avoids the need of this class and associated cascading changes?

Reply:
Yes I think you're on to something there. I will review that tomorrow.

Reply:
OK I've tidied that up now, thanks for the suggestion @ywangd